tests/test_redis_cache.py

```diff
@@ -247,3 +248,98 @@ def __str__(self):
+        mock_client_object.return_value = mock_client
+        redis_cache = RedisCache(self.address, self.port)
+
+        class TestClass(object):
+            def __init__(self):
+                self.call_count = collections.defaultdict(int)
+
+            @redis_cache.cache()
+            def echo(self, parameter):
+                self.call_count[parameter] += 1
+                return parameter
+
+        primitive_argument = 'cache hit test'
+        complex_argument = SimpleObject('test', 42)
+
+        for argument in primitive_argument, complex_argument:
+            # on the first call, the cache is empty
+            instance = TestClass()
+            mock_client.get.return_value = ''
+            instance1_return1 = instance.echo(argument)
+
+            # after the first call, the cache is set, i.e. all the subsequent
+            # calls will return the previously computed value from cache
+            mock_client.get.return_value = pickle.dumps(argument)
+            instance1_return2 = instance.echo(argument)
+            instance1_return3 = instance.echo(argument)
+            self.assertEqual(instance1_return1, argument)
+            self.assertEqual(instance1_return2, argument)
+            self.assertEqual(instance1_return3, argument)
+            self.assertEqual(instance.call_count[argument], 1)
+
+            # the cache should also be hit for calls to the same method on a new
+            # but equivalent object
+            equivalent_instance = TestClass()
+            instance2_return1 = equivalent_instance.echo(argument)
+            instance2_return2 = equivalent_instance.echo(argument)
+            self.assertEqual(instance2_return1, argument)
+            self.assertEqual(instance2_return2, argument)
+            self.assertEqual(equivalent_instance.call_count[argument], 0)
+
+    @patch('redis_cache.redis_cache.RedisClient')
+    def test_cache_on_stateful_class_without_str(self, mock_client_object):
+        """
+        Tests that the cache gets populated when invoking a method on an object
+        that is stateful and that doesn't implement a custom __str__ method.
+
+        Some object are stateful, i.e. their methods will return different
+        values at different times in the object's life-cycle. When the state of
+        the object changes, we need to make sure that the cache doesn't return
+        values for the previous state of the object.
+        """
+        mock_client = Mock()
+        mock_client_object.return_value = mock_client
+        redis_cache = RedisCache(self.address, self.port)
+
+        class TestClass(object):
+            def __init__(self):
+                self.state = None
+
+            @redis_cache.cache()
+            def some_method(self):
+                return self.state
+
+        state1 = 'some simple type state'
+        state2 = SimpleObject('some complex type state', 123)
+
+        def cache_get(cache_key):
+            if cache_key == '-5024234577788600450':
```

Comment:
The cache key formatting is now a little bit more complicated so this required another refactor: rationalize the cache key formatting logic into one place so that it can be re-used from the unit tests.